File: NEETCODE/Trees/invertBinaryTree.py
from typing import Optional
from binarySearchTree import TreeNode
from binarySearchTree import BinarySearchTree


class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def swap_pointers(root):
            if root == None:
                return

            temp = root.left
            root.left = root.right
            root.right = temp

            swap_pointers(root.left)
            swap_pointers(root.right)

        swap_pointers(root)
        return root


if __name__ == "__main__":
    tree = BinarySearchTree()
    tree.insert(5)
    tree.insert(2)
    tree.insert(3)
    tree.insert(6)
    # tree.insert(10)
    # tree.insert(8)
    # tree.insert(20)
    # tree.insert(15)
    # tree.insert(21)

    print('BEFORE:')
    tree.breadthFirstSearch()

    obj = Solution()
    obj.invertTree(tree.root)

    print('AFTER:')
    tree.breadthFirstSearch()

File: C++/Binary Search Tree/invertBinaryTree.cpp
#include <iostream>
#include <algorithm>

#include "binarySearchTree.hpp"

using namespace std;

class Solution
{
private:
    void swapPointers(TreeNode *it)
    {
        if (it == nullptr)
            return;

        // Swapping where left and right point
        TreeNode *temp = it->left;
        it->left = it->right;
        it->right = temp;

        swapPointers(it->left);
        swapPointers(it->right);
    }

public:
    TreeNode *invertTree(TreeNode *root)
    {
        swapPointers(root);

        return root;
    }
};

int main()
{

    BinarySearchTree tree;

    tree.insert(4);
    tree.insert(2);
    tree.insert(7);
    // tree.insert(1);
    // tree.insert(6);

    cout << "Before: ";
    tree.inorder();

    Solution mySol;
    mySol.invertTree(tree.getRoot());

    cout << "After: ";
    tree.inorder();

    return 0;
}